lib/storage/bucket.js

```diff
@@ -349,6 +351,142 @@ Bucket.prototype.setMetadata = function(metadata, callback) {
 };
 
 /**
+ * Make the bucket listing and optionally its contents publicly readable to
+ * anybody.
+ * NOTE: This function may be long-running and use a high number of requests as
+ * it recursively applies makePublic to every file within the bucket.
+ *
+ * @param {object=}  options - The configuration object.
+ * @param {boolean}  options.recursive - Whether to recursively apply makePublic
+ *     to every file in the bucket as well. Default is false.
+ * @param {boolean}  options.force - By default, makePublic will return at the
+ *     first error found, but set this to true to continue processing if this
+ *     occurs.
+ * @param {Function} callback - The callback function.
+ *
+ * @example
+ * bucket.makePublic({ recursive: true }, function(err, files) {
+ *   // err is the first error to occur, otherwise null.
+ *   // files is an array of files successfully made public in the bucket.
+ * });
+ */
+Bucket.prototype.makePublic = function(options, callback) {
+  if (util.is(options, 'function')) {
+    callback = options;
+    options = {};
+  }
+
+  options.public = true;
+  this._makePublicPrivate(options, callback);
+};
+
+Bucket.prototype.makePrivate = function(options, callback) {
+  if (util.is(options, 'function')) {
+    callback = options;
+    options = {};
+  }
+
+  options.private = true;
+  this._makePublicPrivate(options, callback);
+};
+
+Bucket.prototype._makePublicPrivate = function(options, callback) {
+  var self = this;
+  var errors = [];
+  var filesProcessed = [];
+  var nextQuery = {};
+
+  // Allow public to read bucket contents
+  // while preserving original permissions
+  if (options.public) {
+    this.acl.add({
+      entity: 'allUsers',
+      role: 'READER'
+    }, processAllFiles);
+  } else if (options.private) {
+    var query = {
+      predefinedAcl: options.strict ? 'private' : 'projectPrivate'
+    };
+
+    // You aren't allowed to set both predefinedAcl & acl properties on a bucket
+    // so acl must explicitly be nullified.
+    var metadata = { acl: null };
+
+    this.makeReq_('PATCH', '', query, metadata, function(err, resp) {
+      if (err) {
+        processAllFiles(err);
+        return;
+      }
+      self.metadata = resp;
+      processAllFiles(null);
+    });
+  }
+
+  function processAllFiles(err) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    if (options.recursive) {
+      async.whilst(function() {
+        // while there are still more files to make public
+        return !!nextQuery;
+      }, function(cb) {
+        processMoreFiles(nextQuery, cb);
+      }, function(err) {
+        if (err || errors.length) {
+          callback(err || errors, filesProcessed);
+        } else {
+          callback(null, filesProcessed);
+        }
+      });
+    } else {
+      callback(null, filesProcessed);
```

Comment:
I guess the alternative is just doing `bucket.getFiles()` and returning all of the files? But that feels silly. Alternative alternative is documenting that `files` is only populated if you do recursive?